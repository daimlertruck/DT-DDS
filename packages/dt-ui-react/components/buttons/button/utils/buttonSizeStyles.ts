import { CustomTheme as Theme } from '../../../../types';
import { ButtonSizes } from '../constants/types';

export const buttonSizeStyles = (size: ButtonSizes, theme: Theme) =>
  ({
    large: `
      ${theme.fontStyles.button1}
      padding: ${theme.spacing['2xs']} ${theme.spacing.s};
    `,
    medium: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['3xs']} ${theme.spacing.xs};
    `,
    small: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['4xs']} ${theme.spacing['2xs']};
    `,
  }[size]);
