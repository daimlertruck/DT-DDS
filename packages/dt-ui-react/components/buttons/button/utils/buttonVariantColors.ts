import { CustomTheme as Theme } from '../../../../types';
import {
  ButtonVariant,
  ContainedColors,
  OutlinedColors,
  TextColors,
} from '../constants/types';

interface ColorKeys {
  color: string;
  bgColor: string;
  hoverBgColor?: string;
  hoverColor?: string;
}

export const buttonVariantColors = (
  color: ContainedColors | OutlinedColors,
  theme: Theme
): Record<ButtonVariant, ColorKeys> => ({
  contained: {
    color: theme.colors.alpha.white,
    bgColor: theme.palette.alpha[color],
    hoverBgColor: theme.palette.alpha[`${color}Dark_300` as ContainedColors],
  },
  outlined: {
    color:
      color === 'neutral'
        ? theme.colors.alpha.gray_200
        : theme.palette.alpha[color],
    bgColor: theme.colors.alpha.white,
    hoverBgColor: theme.palette.alpha[`${color}Light_300` as OutlinedColors],
  },
  text: {
    color: theme.palette.alpha[color],
    bgColor: 'transparent',
    hoverColor: theme.palette.alpha[`${color as TextColors}Dark_300`],
  },
});
