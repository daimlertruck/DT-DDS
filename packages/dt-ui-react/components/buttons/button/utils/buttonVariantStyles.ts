import { CustomTheme as Theme } from '../../../../types';
import { ButtonVariant, OutlinedColors, SolidColors } from '../constants/types';

export const buttonVariantStyles = (
  variant: ButtonVariant,
  theme: Theme,
  color: SolidColors | OutlinedColors
) =>
  ({
    solid: `
      text-transform: uppercase;
      border: ${theme.spacing.none};

      &:focus {
        box-shadow: 0px 2px 1px 0px #1A1A1A0D, 0px 0px 0px 4px ${theme.colors.blue_10};
        background-color: ${theme.palette[color]};
      }

      &:disabled {
        opacity: 38%;
      }
  `,
    outlined: `
      border: 1px solid;
      text-transform: uppercase;
        ${
          color === 'neutral'
            ? theme.palette.neutralDark_100
            : theme.palette[color]
        };

      &:disabled {
        color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        background-color: ${theme.colors.white};
        border-color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        cursor: default;
      }
    `,
    text: `
      border: none;

      &:disabled {
        color: ${theme.palette.neutralLight_100};
      }
    `,
  }[variant]);
