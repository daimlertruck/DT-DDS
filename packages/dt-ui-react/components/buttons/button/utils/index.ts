export * from './buttonVariantColors';
export * from './buttonVariantStyles';
