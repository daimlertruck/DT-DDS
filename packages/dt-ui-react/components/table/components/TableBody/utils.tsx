import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { TableDataCellHeader } from '../TableDataCell/TableDataCell';

export const addHeadersToDataCells = (
  dataCells: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(dataCells, (dataCell, index) => {
    if (isValidElement(dataCell)) {
      const header = headers[index];
      return cloneElement(
        dataCell,
        {},
        <>
          <TableDataCellHeader>{header}</TableDataCellHeader>
          {dataCell.props.children}
        </>
      );
    }
    return dataCell;
  });
};

export const addHeadersToDataCellsInRows = (
  rows: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(rows, (row) => {
    if (isValidElement(row)) {
      return cloneElement(
        row,
        {},
        addHeadersToDataCells(row.props.children, headers)
      );
    }
    return row;
  });
};
