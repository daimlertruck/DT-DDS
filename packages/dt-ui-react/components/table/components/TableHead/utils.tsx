import { Children, ReactNode, isValidElement } from 'react';

const extractColumnHeaders = (columnHeaders: ReactNode): ReactNode[] => {
  return Children.toArray(columnHeaders).reduce<ReactNode[]>(
    (headers, columnHeader) => {
      if (isValidElement(columnHeader)) {
        return [...headers, columnHeader.props.children];
      }

      return headers;
    },
    []
  );
};

export const extractTableHeaders = (theadRow: ReactNode) => {
  if (isValidElement(theadRow)) {
    return extractColumnHeaders(theadRow.props.children);
  }

  return [];
};
