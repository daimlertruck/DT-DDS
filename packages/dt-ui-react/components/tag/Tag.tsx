import { BaseProps } from '@dt-ui/react-core';

import { CloseIcon } from '../../core/assets';

import { TagBorder, TagColor, TagSize, TagVariant } from './constants';
import { TagStyled, TagGroupStyled, TagButtonCloseStyled } from './Tag.styled';

type TagConditionalProps =
  | {
      onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
      onClose?: never;
    }
  | {
      onClose?: (event: React.MouseEvent<HTMLButtonElement>) => void;
      onClick?: never;
    };

export type TagProps = {
  variant?: TagVariant;
  color?: TagColor;
  border?: TagBorder;
  size?: TagSize;
  isDisabled?: boolean;
  isClickable?: boolean;
} & TagConditionalProps &
  BaseProps;

export const Tag = ({
  style,
  children,
  dataTestId,
  variant = 'solid',
  color = 'primary',
  border = 'squared',
  size = 'small',
  isClickable = false,
  isDisabled = false,
  onClick,
  onClose,
}: TagProps) => {
  const dataTestIdName = dataTestId ?? 'tag';
  const isDismissible = !!onClose;
  const hasHoverStyle =
    !isDisabled && (!!onClick || isClickable || isDismissible);
  const hasClickableStyle = !isDismissible && hasHoverStyle;

  const handleClickTag = (event: React.MouseEvent<HTMLDivElement>) => {
    if (isDisabled || isDismissible) {
      return;
    }

    onClick?.(event);
  };

  return (
    <TagStyled
      aria-disabled={isDisabled}
      border={border}
      color={color}
      data-testid={dataTestIdName}
      hasHover={hasHoverStyle}
      isClickable={hasClickableStyle}
      isDisabled={isDisabled}
      isDismissible={isDismissible}
      onClick={handleClickTag}
      size={size}
      style={style}
      variant={variant}
    >
      {children}

      {isDismissible ? (
        <TagButtonCloseStyled
          aria-disabled={isDisabled}
          color={color}
          data-testid={`${dataTestIdName}-close-button`}
          disabled={isDisabled}
          hasHover={hasHoverStyle}
          onClick={onClose}
          variant={variant}
        >
          <CloseIcon height={12} width={12} />
        </TagButtonCloseStyled>
      ) : null}
    </TagStyled>
  );
};

export const TagGroup = ({ children, style }: BaseProps) => {
  return <TagGroupStyled style={style}>{children}</TagGroupStyled>;
};
