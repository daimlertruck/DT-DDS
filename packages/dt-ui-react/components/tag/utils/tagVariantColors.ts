import { CustomTheme as Theme } from '../../../types';
import { TagColor, TagVariant } from '../constants';

interface TagVariantColors {
  border: string | number;
  color: string;
  borderColor?: string;
  backgroundColor: string;
  '&:hover'?: {
    borderColor?: string;
    backgroundColor: string;
    color: string;
  };
}

interface TagVariantColorsProps {
  theme: Theme;
  color: TagColor;
  variant: TagVariant;
  hasHover: boolean;
  isDisabled: boolean;
}

const getColorWithShade = ({
  theme,
  color,
  shade,
}: Pick<TagVariantColorsProps, 'color'> & {
  shade: '90' | '60' | '30' | '10';
  theme: Theme;
}) => {
  const { colors, palette } = theme;

  if (color === 'primary') {
    return {
      '90': palette.primaryDark_500,
      '60': palette.primaryDark_200,
      '30': palette.primaryLight_100,
      '10': palette.primaryLight_400,
    }[shade];
  }

  const colorWithShade: keyof Theme['colors'] = `${color}_${shade}`;

  return colors[colorWithShade];
};

export const tagVariantColors = ({
  theme,
  color,
  variant,
  hasHover,
  isDisabled,
}: TagVariantColorsProps): TagVariantColors => {
  const colorShade90 = getColorWithShade({
    theme,
    color,
    shade: '90',
  });
  const colorShade60 = getColorWithShade({
    theme,
    color,
    shade: '60',
  });
  const colorShade30 = getColorWithShade({
    theme,
    color,
    shade: '30',
  });
  const colorShade10 = getColorWithShade({
    theme,
    color,
    shade: '10',
  });

  const colorWithDisabled = isDisabled ? colorShade30 : colorShade60;

  return {
    solid: {
      border: theme.spacing.none,
      color: theme.colors.white,
      backgroundColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          backgroundColor: colorShade90,
          color: theme.colors.white,
        },
      }),
    },
    outlined: {
      border: `1px solid ${colorShade60}`,
      color: colorWithDisabled,
      backgroundColor: colorShade10,
      borderColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          borderColor: colorShade90,
          backgroundColor: colorShade10,
          color: colorShade90,
        },
      }),
    },
  }[variant];
};
