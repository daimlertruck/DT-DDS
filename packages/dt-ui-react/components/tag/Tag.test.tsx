import { fireEvent, screen, render } from '@testing-library/react';

import { withProviders } from '../../utils';

import { Tag } from './Tag';

describe('<Tag /> component', () => {
  const ProvidedTag = withProviders(Tag);

  it('should render a span html element with the given content', () => {
    const { container } = render(<ProvidedTag>Active</ProvidedTag>);

    expect(container).toMatchSnapshot();
  });

  it('should show ellipsis if the content exceeds the limit width', async () => {
    const { container } = render(
      <ProvidedTag>This is a very long tag label</ProvidedTag>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render a wrapper for a group of tags', () => {
    const { container } = render(
      <Tag.Group>
        <ProvidedTag>Active</ProvidedTag>
        <ProvidedTag>Inactive</ProvidedTag>
      </Tag.Group>
    );

    expect(container).toMatchSnapshot();
  });

  describe('when the tag is clicked', () => {
    it('should trigger the onClick function', () => {
      const onClick = jest.fn();

      render(<ProvidedTag onClick={onClick}>Click me</ProvidedTag>);

      const tag = screen.getByTestId('tag');
      fireEvent.click(tag);

      expect(onClick).toBeCalledTimes(1);
    });
  });

  describe('Basic tag veriant', () => {
    it('should render a span html element with given styles', () => {
      const { container } = render(
        <ProvidedTag variant='basic'>Active</ProvidedTag>
      );

      expect(container).toMatchSnapshot();
    });

    it('should trigger the onClose function', () => {
      const onClose = jest.fn();

      render(
        <ProvidedTag variant='basic' onClose={onClose} dataTestId='tag'>
          Close me
        </ProvidedTag>
      );

      const closeButton = screen.getByTestId('tag-close-button');
      fireEvent.click(closeButton);

      expect(onClose).toBeCalledTimes(1);
    });

    describe('when the tag is clickable', () => {
      it('should trigger the onClick function', () => {
        const onClick = jest.fn();

        render(
          <ProvidedTag variant='basic' onClick={onClick} isClickable={true}>
            Click me
          </ProvidedTag>
        );

        const tag = screen.getByTestId('tag');
        fireEvent.click(tag);

        expect(onClick).toBeCalledTimes(1);
      });
    });
  });
});
