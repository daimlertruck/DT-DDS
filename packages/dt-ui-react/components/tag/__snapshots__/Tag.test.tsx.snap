// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> component should render a wrapper for a group of tags 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 8px;
}

.emotion-2 {
  font-size: 0.625rem;
  font-weight: 700;
  line-height: 0.875rem;
  padding: 2px 8px;
  height: 18px;
  border: 0;
  color: #FFFFFF;
  background-color: #013C6B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      aria-disabled="false"
      class="emotion-2 emotion-3"
      color="primary"
      data-testid="tag"
    >
      Active
    </div>
    <div
      aria-disabled="false"
      class="emotion-2 emotion-3"
      color="primary"
      data-testid="tag"
    >
      Inactive
    </div>
  </div>
</div>
`;

exports[`<Tag /> component should render the tag with variant outlined, color blue, border squared, size small, isClickable as true, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.625rem;
  font-weight: 700;
  line-height: 0.875rem;
  padding: 2px 8px;
  height: 18px;
  border: 1px solid #008DFC;
  color: #008DFC;
  background-color: #E6F5FF;
  border-color: #008DFC;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 0px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.emotion-0:hover {
  border-color: #013C6B;
  background-color: #E6F5FF;
  color: #013C6B;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="blue"
  data-testid="tag"
>
  Click me
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color grey, border rounded, size medium, isClickable as false, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0;
  color: #FFFFFF;
  background-color: #7C7C7D;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.emotion-0:hover {
  background-color: #323232;
  color: #FFFFFF;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="grey"
  data-testid="tag"
>
  Click me
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color grey, border rounded, size medium, isClickable as false, isDisabled as false, onClick and onClose 2`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0;
  color: #FFFFFF;
  background-color: #7C7C7D;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0:hover {
  background-color: #323232;
  color: #FFFFFF;
}

.emotion-2 {
  background-color: transparent;
  color: currentColor;
  display: inherit;
  border: 0;
  cursor: pointer;
  border-radius: 16px;
}

.emotion-2:hover {
  background-color: #FFFFFF;
  color: #323232;
}

.emotion-2:disabled {
  cursor: not-allowed;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="grey"
  data-testid="tag"
>
  Click me
  <button
    aria-disabled="false"
    class="emotion-2 emotion-3"
    color="grey"
    data-testid="tag-close-button"
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color grey, border rounded, size medium, isClickable as false, isDisabled as true, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0;
  color: #FFFFFF;
  background-color: #D3D4D5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: not-allowed;
}

.emotion-2 {
  background-color: transparent;
  color: currentColor;
  display: inherit;
  border: 0;
  cursor: pointer;
  border-radius: 16px;
}

.emotion-2:disabled {
  cursor: not-allowed;
}

<div
  aria-disabled="true"
  class="emotion-0 emotion-1"
  color="grey"
  data-testid="tag"
>
  Click me
  <button
    aria-disabled="true"
    class="emotion-2 emotion-3"
    color="grey"
    data-testid="tag-close-button"
    disabled=""
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`<Tag /> component should render the tag with variant solid, color primary, border rounded, size medium, isClickable as false, isDisabled as false, onClick and onClose 1`] = `
.emotion-0 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  padding: 8px 12px;
  height: 32px;
  border: 0;
  color: #FFFFFF;
  background-color: #013C6B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-transform: uppercase;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  gap: 4px;
  border-radius: 32px;
}

<div
  aria-disabled="false"
  class="emotion-0 emotion-1"
  color="primary"
  data-testid="tag"
>
  Click me
</div>
`;
